### PR TITLE
DAOS-10342 test: refine hard_rebuild.py #8743

### PR DESCRIPTION
**Pull request by liuxuezhao**
in hard_rebuild.py, remove the first time ior_read_dataset() calling,
as it will zero the stoneWallingStatusFile's content and fail
the 2nd ior_read_dataset.

Test-tag: pr ec_ior_hard_online_rebuild

Signed-off-by: Xuezhao Liu <xuezhao.liu@intel.com>

### CODE REVIEW

**Review by daltonbohning — COMMENTED** [review]
> hard_rebuild.yaml, add "-c" (collective) flag for ior write to
   avoid generate holes (see ior code WriteOrRead() in ior.c,
   hitStonewall related code).

This seems like more of a workaround than a fix. Shouldn't rebuild still work with holes? It's possible for files in a production system to have holes.

**Comment by liuxuezhao** [comment]
> > hard_rebuild.yaml, add "-c" (collective) flag for ior write to
> > avoid generate holes (see ior code WriteOrRead() in ior.c,
> > hitStonewall related code).
> 
> This seems like more of a workaround than a fix. Shouldn't rebuild still work with holes? It's possible for files in a production system to have holes.

the real fix is remove the first ior read, please check details in the ticket. "-c" is to avoid read hole but is not really must needed.

**Review by daltonbohning — DISMISSED** [review]

**Comment by mchaarawi** [comment]
> in IOR code, in stoneWalling mode testing, without -c possibly leave hole in the file.
> but that should not cause the EOF error. I am fine to remove it.

yea what this does is that it just coordinates ranks in case of stonewalling to exit together rather than every rank to check for stonewalling timer on their own. so yea it can cause some ranks to write more. but it has no affect whatsover on the transfer size or IO size that DAOS uses if you are not using -a MPIIO.
so it really should not have any affect on the expectation of the IO.

**Comment by daltonbohning** [comment]
> in IOR code, in stoneWalling mode testing, without -c possibly leave hole in the file. but that should not cause the EOF error. I am fine to remove it.

We should remove it, but since this is a cherry-pick, I don't think we should remove it here, since that would leave master and release/2.2 out of sync. Which would create conflicts if another commit modified the same area. It's better for the cherry pick to be exact.

I have a ticket for scrubbing the ftest/ior directory where I can handle this (I left a TODO that I can handle): https://daosio.atlassian.net/browse/DAOS-9944

**Review by daltonbohning — APPROVED** [review]

**Comment by liuxuezhao** [comment]
> 1. stoneWallingStatusFile

@daltonbohning @mchaarawi 
hmm, I checked again the ior code (WriteOrRead() in ior.c, hitStonewall related code), we cannot remove the “-c” option.
because the “hitStonewall” possibly be different for different ranks, then that rank will not write that part of data to the shared file. But IOR will write the max “point->pairs_accessed” to the file of stoneWallingStatusFile and use that value to read for following ior_read. If the last IOR rank happened to with less “point->pairs_accessed” then during ior_write it possibly will write less length to the file (“offset += (i * test->numTasks * test->blockSize) + (pretendRank * test->blockSize)”), then the ior_read possibly will hit EOF prematurely failure as it read beyond EOF.
this has nothing to do with rebuild.

**Comment by mchaarawi** [comment]
> > 1. stoneWallingStatusFile
> 
> @daltonbohning @mchaarawi hmm, I checked again the ior code (WriteOrRead() in ior.c, hitStonewall related code), we cannot remove the “-c” option. because the “hitStonewall” possibly be different for different ranks, then that rank will not write that part of data to the shared file. But IOR will write the max “point->pairs_accessed” to the file of stoneWallingStatusFile and use that value to read for following ior_read. If the last IOR rank happened to with less “point->pairs_accessed” then during ior_write it possibly will write less length to the file (“offset += (i * test->numTasks * test->blockSize) + (pretendRank * test->blockSize)”), then the ior_read possibly will hit EOF prematurely failure as it read beyond EOF. this has nothing to do with rebuild.

there could be an issue here, so i can look more tomorrow. but just to be clear, we always run ior-hard without -c and with stonewalling and we never had an issue before. so i doubt it's true that -c is required and the IOR maintainers have not seen and fixed that issue.

**Comment by daltonbohning** [comment]
> there could be an issue here, so i can look more tomorrow. but just to be clear, we always run ior-hard without -c and with stonewalling and we never had an issue before. so i doubt it's true that -c is required and the IOR maintainers have not seen and fixed that issue.

Right. IO500, Frontera, and some CI tests run without `-c`. But removing `-c` would need a followup PR anyway, so I don't think we need to block this one. I can also investigate if this is a test issue as I scrub the ftest/ior directory. In general, this test case is very confusing/complex to follow compared to most other IOR tests.

**Comment by mchaarawi** [comment]
> hmm, I checked again the ior code (WriteOrRead() in ior.c, hitStonewall related code), we cannot remove the “-c” option.
> because the “hitStonewall” possibly be different for different ranks, then that rank will not write that part of data to the shared file. But IOR will write the max “point->pairs_accessed” to the file of stoneWallingStatusFile and use that value to read for following ior_read. If the last IOR rank happened to with less “point->pairs_accessed” then during ior_write it possibly will write less length to the file (“offset += (i * test->numTasks * test->blockSize) + (pretendRank * test->blockSize)”), then the ior_read possibly will hit EOF prematurely failure as it read beyond EOF.
> this has nothing to do with rebuild.

if you check after the stonewall loop has exit, there is a check for each rank to see if it has written that many pairs, and if it doesn't it continues writing till it writes that many pairs:
```if(pairCnt != point->pairs_accessed){
            // some work needs still to be done, complete the current block !  ```

**Comment by mchaarawi** [comment]
> so I don't think we need to block this one

im not so sure we should just land this without being able to understand the issue more especially since this is going on the release branch. i don't know what this test is trying to do with rebuild. I have not looked at the test script but was just chiming in on the question with -c.  ill ping Dalton tomorrow to understand. maybe the test steps itself we have here requires to have -c, then this is less of an issue, but i'd like to understand if that is the case. whether this is a rebuild issue or not is still not clear to me TBH.

**Comment by liuxuezhao** [comment]
> if(pairCnt != point->pairs_accessed){
>             // some work needs still to be done, complete the current block ! 
>  ...
> ```
Ah, OK, I think you are right, the "-c" is not needed.

**Comment by liuxuezhao** [comment]
> im not so sure we should just land this without being able to understand the issue more especially since this is going on the release branch. i don't know what this test is trying to do with rebuild. I have not looked at the test script but was just chiming in on the question with -c. ill ping Dalton tomorrow to understand. maybe the test steps itself we have here requires to have -c, then this is less of an issue, but i'd like to understand if that is the case. whether this is a rebuild issue or not is still not clear to me TBH.

well, the real fix is remove the 1st ior_read, details explained in the ticket. For the specific ticket should not a rebuild issue.

**Comment by liuxuezhao** [comment]
> Ah, OK, I think you are right, the "-c" is not needed.
just removed the -c option for this patch, I'll push another patch to remove that option in master too.

**Comment by mchaarawi** [comment]
> well, the real fix is remove the 1st ior_read, details explained in the ticket. For the specific ticket should not a rebuild issue.

yes im good with that one, my comment on not landing was just for -c

**Review by mchaarawi — APPROVED** [review]

**Review by daltonbohning — APPROVED** [review]

**Pull request merged**
